src/chainstate/stacks/miner.rs

```diff
@@ -356,29 +616,37 @@ impl<'a> StacksMicroblockBuilder<'a> {
                 "Adding microblock tx {} would exceed epoch data size",
                 &tx.txid()
             );
-            return Ok(None);
+            return Ok(TransactionResult::skipped(
+                &tx,
+                "Adding microblock tx would exceed epoch data size".to_string(),
+                true,
+            ));
         }
         let quiet = !cfg!(test);
         match StacksChainState::process_transaction(clarity_tx, &tx, quiet) {
-            Ok((_, receipt)) => return Ok(Some(receipt)),
-            Err(e) => match e {
-                Error::CostOverflowError(cost_before, cost_after, total_budget) => {
-                    // note: this path _does_ not perform the tx block budget % heuristic,
-                    //  because this code path is not directly called with a mempool handle.
-                    warn!(
-                        "Transaction {} reached block cost {}; budget was {}",
-                        tx.txid(),
-                        &cost_after,
-                        &total_budget
-                    );
-                    clarity_tx.reset_cost(cost_before);
-                }
-                _ => {
-                    warn!("Error processing TX {}: {}", tx.txid(), e);
+            Ok((fee, receipt)) => Ok(TransactionResult::success(&tx, fee, receipt)),
+            Err(e) => {
+                match &e {
+                    Error::CostOverflowError(cost_before, cost_after, total_budget) => {
+                        // note: this path _does_ not perform the tx block budget % heuristic,
+                        //  because this code path is not directly called with a mempool handle.
+                        warn!(
+                            "Transaction {} reached block cost {}; budget was {}",
+                            tx.txid(),
+                            &cost_after,
+                            &total_budget
+                        );
+                        clarity_tx.reset_cost(cost_before.clone());
+                        Ok(TransactionResult::error(&tx, e))
```

Comment:
This shouldn't be a `TransactionResult::error`, but instead a `TransactionResult::skipped` with the budget message. The transaction itself didn't necessarily do anything to create this error, rather the block is just full.